### ISSUE TITLE
A performance issue vs pyahocorasick

### ISSUE THREAD

**Issue by hscspring** [BurntSushi/aho-corasick#128]
Hi there,

I used this library to optimize my project, however I got a poorer result vs pyahocorasick (about 3 times slower). 
Would anyone please check the comparison?

My test code: 

https://github.com/hscspring/bytepiece-rs/tree/main/bench

The result:

```bash
# pyahocorasick
Count: 100, Elapsed time: 0.08392333984375 ms
Count: 1000, Elapsed time: 0.7140636444091797 ms
Count: 10000, Elapsed time: 6.103992462158203 ms
Count: 100000, Elapsed time: 61.48505210876465 ms

# aho-corasick
Count: 100, Elapsed time: 301.292µs
Count: 1000, Elapsed time: 2.283459ms
Count: 10000, Elapsed time: 19.507958ms
Count: 100000, Elapsed time: 188.772959ms
```

Environment:

```bash
Python3.9
Mac M2 16G memory

rustup show
Default host: x86_64-apple-darwin
rustup home:  /Users/Yam/.rustup

installed toolchains
--------------------

stable-x86_64-apple-darwin (default)
nightly-2022-04-11-x86_64-apple-darwin
nightly-2022-08-05-x86_64-apple-darwin
nightly-x86_64-apple-darwin
1.51.0-x86_64-apple-darwin

installed targets for active toolchain
--------------------------------------

aarch64-apple-darwin
riscv64gc-unknown-none-elf
wasm32-unknown-unknown
wasm32-wasi
x86_64-apple-darwin

active toolchain
----------------

stable-x86_64-apple-darwin (default)
rustc 1.69.0 (84c898d65 2023-04-16)
```

**Comment by BurntSushi** [comment]
Thank you for the good reproduction. Although your Python program uses `bytes` and not `string` patterns, and this requires providing different compilation options to `pyahocorasick` that I didn't know about. If your reproduction requires a non-default configuration, you should include those details. (And it would be even better to provide a complete set of commands to do any setup or running of your benchmark, because you shouldn't assume maintainers of Rust libraries know Python.) In my case, I had to do this to install `pyahocorasick`:

```
$ AHOCORASICK_BYTES=1 pip install pyahocorasick --no-binary ':all:'
```

The `AHOCORASICK_BYTES=1` tells `pyahocorasick` to compile itself without Unicode support (which is apparently necessary in order to be able to add `bytes` patterns), and the `--no-binary ':all:'` setting is required to force `pip` to compile the C parts of the `pyahocorasick` library. Otherwise, the `AHOCORASICK_BYTES` configuration won't get picked up.

I also changed your programs to print the `count` that you were keeping track of but didn't actually use.

Once I got everything setup, I was able to run the programs:

```
$ cargo build --release

$ time ./target/release/bench
Count: 100, Actual: 398, Elapsed time: 15.659µs
Count: 1000, Actual: 6377, Elapsed time: 138.706µs
Count: 10000, Actual: 66210, Elapsed time: 956.291µs
Count: 100000, Actual: 665241, Elapsed time: 7.898878ms

real    0.947
user    0.290
sys     0.657
maxmem  77 MB
faults  0

$ time python main.py
Count: 100, Actual: 401, Elapsed time: 0.026702880859375 ms
Count: 1000, Actual: 6449, Elapsed time: 0.31566619873046875 ms
Count: 10000, Actual: 67243, Elapsed time: 2.599000930786133 ms
Count: 100000, Actual: 675759, Elapsed time: 25.798559188842773 ms

real    0.122
user    0.104
sys     0.017
maxmem  61 MB
faults  0
```

... and it looks like this library is quite a bit faster than `pyahocorasick`, in direct contradiction to your timings. Now I wonder whether you didn't compile in release mode. So let's try compiling in debug mode and see what happens:

```
$ cargo build

$ time ../target/debug/bench
Count: 100, Actual: 398, Elapsed time: 62.679µs
Count: 1000, Actual: 6377, Elapsed time: 672.36µs
Count: 10000, Actual: 66210, Elapsed time: 5.621455ms
Count: 100000, Actual: 665241, Elapsed time: 55.26504ms

real    1.801
user    1.100
sys     0.700
maxmem  78 MB
faults  0
```

Ah, that appears to be the issue here. It's now about twice as slow as `pyahocorasick`, which is closer to the timings your report. Perhaps you compiled your Rust program in debug mode instead of release mode? Debug mode doesn't enable any optimizations and thus there should be no expectation that the resulting program will be fast. In some cases, it can be extraordinarily slow.

I note that I cannot tell how you built your program because you did not include a complete reproduction.

One other thing worth pointing out here is that the actual counts produced by each program are not equivalent. That could be due to changes in how normalization is being done or something else, or maybe even a bug somewhere. I didn't have time to do a full comparison with the Python program, but I did add to your program a method for counting all overlapping matches using naive single-substring search. The counts match what this crate provides, so I'm confident there are no bugs on my end I think:

```rust
use std::collections::HashMap;
use std::fs::File;
use std::io::Read;
use std::time::Instant;

use aho_corasick::AhoCorasick;
use base64::{engine::general_purpose, Engine as _};
use bytes::Bytes;
use serde::Deserialize;
use unic_normal::StrNormalForm;

#[derive(Debug, Deserialize)]
struct ModelData {
    _id: usize,
    _value: String,
    _freq: usize,
}

fn main() {
    let current_dir = std::env::current_dir().unwrap();
    let root = current_dir.parent().unwrap();

    let model_path = root.join("src/tokenizer/model/bytepiece_80k.model");
    let file = File::open(model_path).unwrap();
    let json: serde_json::Value = serde_json::from_reader(file).unwrap();
    let model: HashMap<String, ModelData> =
        serde_json::from_str(json.to_string().as_str()).unwrap();
    let mut patterns: Vec<Bytes> = Vec::new();
    for (key, _value) in model {
        let token_u8 = general_purpose::STANDARD.decode(key.as_str()).unwrap();
        let token = Bytes::from(token_u8);
        patterns.push(token.clone());
    }
    patterns.sort();
    let aho = AhoCorasick::new(&patterns).unwrap();
    let file_path = current_dir.join("data/鲁迅全集.txt");
    let mut file = File::open(file_path).unwrap();
    let mut text = String::new();
    file.read_to_string(&mut text).unwrap();
    let normalized_text = text.nfc().collect::<String>();
    for total in [100, 1000, 10000, 100000] {
        let test_text =
            normalized_text.chars().take(total).collect::<String>();
        let byte_text = Bytes::from(test_text);
        let mut count = 0;
        let start_time = Instant::now();
        for _ in aho.find_overlapping_iter(byte_text.as_ref()) {
            count += 1;
        }
        let end_time = Instant::now();
        let elapsed_time = end_time - start_time;
        println!(
            "Count: {}, Actual: {}, Elapsed time: {:?}",
            total, count, elapsed_time
        );
    }

    for total in [100, 1000, 10000, 100000] {
        let test_text =
            normalized_text.chars().take(total).collect::<String>();
        let byte_text = Bytes::from(test_text);
        let mut count = 0;
        let start_time = Instant::now();
        for p in patterns.iter() {
            assert!(!p.is_empty());
            let mut haystack = byte_text.as_ref();
            loop {
                let offset = match memchr::memmem::find(haystack, &p) {
                    None => break,
                    Some(offset) => offset,
                };
                count += 1;
                haystack = &haystack[offset + 1..];
            }
        }
        let end_time = Instant::now();
        let elapsed_time = end_time - start_time;
        println!(
            "Count: {}, Actual: {}, Elapsed time: {:?}",
            total, count, elapsed_time
        );
    }
}
```

So bottom line here is that I think you are building your Rust program in debug mode and that is likely the source of the slowness here. If I'm wrong about that, then please provide more detailed reproduction steps.

**Comment by itamarst** [comment]
Also worth noting there are Python bindings to this Rust library (https://github.com/G-Research/ahocorasick_rs/). It doesn't support bytes yet, but it something that could be added.

**Comment by hscspring** [comment]
@BurntSushi Thanks very much for your very detailed reply. 

Actually, it's my own mistake. I didn't realize the issue (debug mode) at that time..

And also didn't provide additional info for the pyahocorasick installation.

@itamarst Thanks for your reply, I know the library, and also have used it in my projects. As you say, it doesn't support bytes yet.
